File: lab/python/textos/placa_de_carro_valido.py
import re

# Antiga: LLL-NNNN (hífen opcional)
RE_ANTIGA = re.compile(r"^[A-Z]{3}-?\d{4}$")

# Mercosul: LLLNLNN (ex.: ABC1D23)
RE_MERCOSUL = re.compile(r"^[A-Z]{3}\d[A-Z]\d{2}$")


def validar_placa_br(placa: str):
    """
    Retorna (True, 'ANTIGA'|'MERCOSUL') se válida, senão (False, None).
    Aceita letras minúsculas e ignora espaços nas pontas.
    """
    p = placa.strip().upper()

    if RE_ANTIGA.match(p):
        return True, "ANTIGA"
    if RE_MERCOSUL.match(p):
        return True, "MERCOSUL"
    return False, ""


# Testes
placas = [
    "ABC-1234",   # antiga (com hífen)
    "ABC1234",    # antiga (sem hífen)
    "BRA2E19",    # mercosul
    "abc1d23",    # mercosul (minúsculas)
    "AB-1234",    # inválida
    "ABCD-1234",  # inválida
    "ABC12D3",    # inválida (ordem errada)
]

for p in placas:
    ok, tipo = validar_placa_br(p)
    print(f"{p:10} -> {'Válida: ' + tipo if ok else 'Inválida'}")

File: lab/javascript/textos/placa_de_carro_valido.js
// Regex antigas e novas (Mercosul)
const RE_ANTIGA = /^[A-Z]{3}-?\d{4}$/;
const RE_MERCOSUL = /^[A-Z]{3}\d[A-Z]\d{2}$/;

function validar_placa_br(placa) {
  /**
   * Retorna [true, 'ANTIGA'|'MERCOSUL'] se válida, senão [false, ''].
   * Aceita letras minúsculas e ignora espaços nas pontas.
   */
  const p = placa.trim().toUpperCase();

  if (RE_ANTIGA.test(p)) return [true, "ANTIGA"];
  if (RE_MERCOSUL.test(p)) return [true, "MERCOSUL"];
  return [false, ""];
}

// -------------------- Testes --------------------
const placas = [
  "ABC-1234",   // antiga (com hífen)
  "ABC1234",    // antiga (sem hífen)
  "BRA2E19",    // mercosul
  "abc1d23",    // mercosul (minúsculas)
  "AB-1234",    // inválida
  "ABCD-1234",  // inválida
  "ABC12D3",    // inválida (ordem errada)
];

for (const p of placas) {
  const [ok, tipo] = validar_placa_br(p);
  console.log(`${p.padEnd(10)} -> ${ok ? "Válida: " + tipo : "Inválida"}`);
}
